drasyl-core/src/main/java/org/drasyl/handler/pubsub/PubSubBrokerHandler.java

```diff
@@ -0,0 +1,129 @@
+    private static final Logger LOG = LoggerFactory.getLogger(PubSubBrokerHandler.class);
+    private final Multimap<String, DrasylAddress> subscriptions;
+
+    PubSubBrokerHandler(final Multimap<String, DrasylAddress> subscriptions) {
+        this.subscriptions = requireNonNull(subscriptions);
+    }
+
+    public PubSubBrokerHandler() {
+        this(new HashSetMultimap<>());
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) {
+        // inform all subscribers that we're shutting down
+        for (final String topic : subscriptions.keySet()) {
+            Collection<DrasylAddress> subscribers = subscriptions.get(topic);
+            for (final DrasylAddress subscriber : subscribers) {
+                ctx.writeAndFlush(new OverlayAddressedMessage<>(PubSubUnsubscribe.of(topic), subscriber));
+            }
+        }
+        if (!subscriptions.isEmpty()) {
+            ctx.flush();
+            subscriptions.clear();
+        }
+
+        ctx.fireChannelInactive();
+    }
+
+    @Override
+    public boolean acceptInboundMessage(final Object msg) throws Exception {
+        return msg instanceof OverlayAddressedMessage && (((OverlayAddressedMessage<?>) msg).content() instanceof PubSubPublish || ((OverlayAddressedMessage<?>) msg).content() instanceof PubSubSubscribe || ((OverlayAddressedMessage<?>) msg).content() instanceof PubSubUnsubscribe);
+    }
+
+    @Override
+    protected void channelRead0(ChannelHandlerContext ctx,
+                                OverlayAddressedMessage<PubSubMessage> msg) throws Exception {
+        LOG.trace("Got `{}` from `{}`.", msg.content(), msg.sender());
+        if (msg.content() instanceof PubSubPublish) {
+            handlePublish(ctx, (PubSubPublish) msg.content(), msg.sender());
+        }
+        else if (msg.content() instanceof PubSubSubscribe) {
+            handleSubscribe(ctx, (PubSubSubscribe) msg.content(), msg.sender());
+        }
+        else if (msg.content() instanceof PubSubUnsubscribe) {
+            handleUnsubscribe(ctx, (PubSubUnsubscribe) msg.content(), msg.sender());
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private void handlePublish(final ChannelHandlerContext ctx,
+                               final PubSubPublish msg,
+                               final DrasylAddress sender) {
+        final Collection<DrasylAddress> subscribers = subscriptions.get(msg.getTopic());
+        if (subscribers.isEmpty()) {
+            LOG.debug("Topic `{}` got new publication from `{}`. But as there are no subscribers, publication is dropped.", msg::getTopic, () -> sender);
+        }
+        else {
+            LOG.debug("Topic `{}` got new publication from `{}`. Forward to {} subscriber(s).", msg::getTopic, () -> sender, subscribers::size);
+            for (final DrasylAddress subscriber : subscribers) {
+                ctx.writeAndFlush(new OverlayAddressedMessage<>(msg.retain(), subscriber));
```

Comment:
Should we try to avoid retain and all the dropped methods in netty5?

Reply:
How do we achieve this?

Reply:
Create a copy and free the old buffer?

Reply:
I think we should wait with such preparation work until the netty 5 API is stable.

Hopefully netty will then provide more mature guidelines.